cql3/restrictions/statement_restrictions.cc

```diff
@@ -145,27 +145,55 @@ to_column_definition(const schema_ptr& schema, const ::shared_ptr<column_identif
 }
 #endif
 
+template <typename Visitor>
+concept visitor_with_binary_operator_context = requires (Visitor v) {
+    { v.current_binary_operator } -> std::same_as<const expr::binary_operator*>;
+};
+
+void with_current_binary_operator(
+        visitor_with_binary_operator_context auto& visitor,
+        std::invocable<const expr::binary_operator&> auto func) {
+    if (!visitor.current_binary_operator) {
+        throw std::logic_error("Evaluation expected within binary operator");
+    }
+    func(*visitor.current_binary_operator);
+}
+
 /// Every token, or if no tokens, an EQ/IN of every single PK column.
 static std::vector<expr::expression> extract_partition_range(
         const expr::expression& where_clause, schema_ptr schema) {
     using namespace expr;
     struct {
         std::optional<expression> tokens;
         std::unordered_map<const column_definition*, expression> single_column;
+        const binary_operator* current_binary_operator = nullptr;
 
         void operator()(const conjunction& c) {
             std::ranges::for_each(c.children, [this] (const expression& child) { std::visit(*this, child); });
         }
 
         void operator()(const binary_operator& b) {
-            if (std::holds_alternative<token>(b.lhs)) {
+            if (current_binary_operator) {
+                throw std::logic_error("Nested binary operators are not supported");
```

Comment:
An special reason why? (or just that you didn't want to implement it yet)

Reply:
There's no way to create a nested binary expression from the parser, so even if I implemented it, it would never get exercised.

I'd like to generalize the parser in the future so that this is possible.

Reply:
(i.e. I expect never to see this thrown).